indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/Asn1Encodable.kt

```diff
@@ -43,88 +43,90 @@ interface Asn1Encodable<A : Asn1Element> {
+     * If the provided [template]'s tagClass is not set, the class of the resulting structure defaults to [TagClass.CONTEXT_SPECIFIC].
+     * If a CONSTRUCTED Tag is applied to an ASN.1 Primitive, the CONSTRUCTED bit is overridden and set to zero
      */
-    fun decodeFromTlvSafe(src: A) = catching { decodeFromTlv(src) }
+    infix fun withImplicitTag(template: Tag.Template) = encodeToTlv().withImplicitTag(template)
+}
 
+/**
+ * Interface providing convenience methods to decode from ASN.1.
+ * Especially useful when companion objects of classes implementing [Asn1Encodable] implement it.
+ */
+interface Asn1Decodable<A : Asn1Element, T : Asn1Encodable<A>> {
     /**
-     * Convenience method, directly DER-decoding a byte array to [T]
-     * @throws Asn1Exception if invalid data is provided
+     * Processes an [A], parsing it into an instance of [T]
+     * @throws Asn1Exception if invalid data is provided.
+     * Allows overriding the tag, if it deviates from the element's default tag.
+     * Specify [tagOverride] for verifying implicitly tagged elements (and better not override this function).
+     * @throws Asn1Exception
      */
     @Throws(Asn1Exception::class)
-    fun decodeFromDer(src: ByteArray): T = decodeFromTlv(Asn1Element.parse(src) as A)
+    fun decodeFromTlv(src: A, tagOverride: Asn1Element.Tag? = null): T {
+        verifyTag(src, tagOverride)
+        return doDecode(src)
+    }
 
     /**
-     * Exception-free version of [decodeFromDer]
-     */
-    fun decodeFromDerOrNull(src: ByteArray) = catching { decodeFromDer(src) }.getOrNull()
-
-    /**
-     * Safe version of [decodeFromDer], wrapping the result into a [KmmResult]
-     */
-    fun decodeFromDerSafe(src: ByteArray) = catching { decodeFromDer(src) }
-}
-
-interface Asn1TagVerifyingDecodable<T : Asn1Encodable<Asn1Primitive>> :
-    Asn1Decodable<Asn1Primitive, T> {
-
-    /**
-     * Same as [Asn1Decodable.decodeFromTlv], but allows overriding the tag, should the implementing class verify it.
-     * Useful for implicit tagging, in which case you will want to call [at.asitplus.signum.indispensable.asn1.DERTags.toImplicitTag] on [tagOverride].
+     * Actual element-specific decoding function. By default, this is invoked after [verifyTag]
      * @throws Asn1Exception
      */
     @Throws(Asn1Exception::class)
-    fun decodeFromTlv(src: Asn1Primitive, tagOverride: Asn1Element.Tag?): T
+    fun doDecode(src: A): T
+
+    @Throws(Asn1TagMismatchException::class)
+    fun verifyTag(src: A, tagOverride: Asn1Element.Tag?) {
+        val expected = tagOverride ?: return
+        if (src.tag != expected)
+            throw Asn1TagMismatchException(expected, src.tag)
+    }
 
     /**
      * Exception-free version of [decodeFromTlv]
      */
-    fun decodeFromTlvOrNull(src: Asn1Primitive, tagOverride: Asn1Element.Tag?) =
+    fun decodeFromTlvOrNull(src: A, tagOverride: Asn1Element.Tag? = null) =
```

Comment:
wasn't the plan to get rid of all of the tag override parameters? do we still need these?

Reply:
??? you absolutely need this to decode while verifying implicit tags.

Reply:
ah, it's not really a "tag override", it's more of an "asserted tag"? is this the right name?

Reply:
you are right. it only checks if present. fixed!